### PR TITLE
20210709 pr heli yaw check fix

### PR DESCRIPTION
**Pull request by lthall**
This removes the reset of yaw rate when disarmed to let heli users check yaw control on the ground before flight.
Ironically Heli doesn't do this in their own dedicated acro mode. This makes me wonder why we are compromising more semi autonomous modes when heli didn't think it was important enough in the most manual input mode....

I have removed the reset of yaw rate during all phases of Stabilize_Heli.

I have not removed it in any aspect of Loiter as you can't check roll and pitch in this mode either.

My personal opinion is that any checks should be done in Stabilize_Heli where you can check roll, pitch, yaw and collective (or acro if they allowed it).

To summarize the behaviour:
Heli Control checks (via the rate controllers) can be run on all four axis in Acro and Stab when in Motor_Stopped and Landed_Ground_Idle .
Heli Control checks (via the rate controllers) can be run on the roll, pitch and yaw axis in non-GPS, altitude hold based modes when in Motor_Stopped.
Other actuator frame types provide direct actuator output when disarmed in all modes OR control checks (via the rate controllers) when in Motor_Stopped in non-GPS modes.

### CODE REVIEW

**Comment by Hwurzburg** [comment]
I dont see why this would not be added to heli ACRO mode also...

**Review by Gone4Dirt — COMMENTED** [review]
Thanks for looking at this @lthall
I backported these changes to 4.1 beta and have tested IRL.  The changes you made fix the issue in the flight modes you made the changes.
I agree with Henry we should allow the control checks in acro.  I also noticed we are calling reset_target_and_rate() in acro so we don't have control output in pitch or roll either.  I think we should change this reset_yaw_target_and_rate(false) in acro.
You have only applied the change to the MotorStopped  case in these modes.  I think we should also allow control checks in the Landed_Ground_Idle case so as to let those with IC engines do there checks with engine running at idle.  Was there a particular case you are trying to avoid by not putting tchange in Ground_Idle?
With this change, we are allowing control checks to be done in PosHold but not in Loiter.  Is there a reason that you have not added it to Loiter?  IMHO I think we should also alloy checks in Loiter if we are doing them in Pos_Hold so that the two position control modes (whereby pilots are still inputing something) at least exhibit the same behaviour.  Also, I think i am correct in saying QLoiter is probablly the most commonly ised VTOL mode, and allowing the control checks on the ground in loiter should read across to QLoiter (I think anyway...).

**Comment by lthall** [comment]
>> I don't see why this would not be added to Heli ACRO mode also

This is a pre existing heli issue. I am not the Heli maintainer so I will need to defer to Bill. @bnsgeyer I can chat to you about this if you like.

>> You have only applied the change to the MotorStopped case in these modes.
Yes, it is hard to justify doing this at all but we can do this when MotorStopped without risk to aircraft. I don't see any justification to doing this in Landed_Ground_Idle when they can do it before in MotorStopped.

Landed_Ground_Idle is where we prepare the aircraft for takeoff. The presence of yaw rate from arming the aircraft can cause the aircraft to rotate on the ground, this prevents this from happening. It was the absence of this that resulted in a 50 kg multirotor rotating on the ground, wrap up a line into the propellers, and flip over while following normal take off procedures. While I understand the convenience of being able "to let those with IC engines do there checks with engine running at idle" but this has to be weighed against the inconvenience of losing an aircraft or worse, having a person injured.

>> PosHold but not in Loiter.

PosHold has direct manual control of roll and pitch. In Loiter you do not have direct manual control of roll and pitch therefore the only axis you can check is Yaw. We should probably double check that you can exercise roll and pitch in Pos_Hold after arming and if we can't we should remove it from Pos_Hold.

My personal opinion is that I can't justify allowing these checks in any mode that does not provide direct control over roll, pitch and yaw. If the pilot is checking the aircraft is flight worthy then they should be using stabilize where they can also exercise collective and in that case I can only justify doing this in Stab and acro and we should not provide it in Alt_Hold or Pos_Hold either.

>>  that should read across to QLoiter (I think anyway...) to aid with vectored yaw VTOL planes, as Tridge pointed out.
We aren't doing it in Loiter and there is no reason these axis can't be checked using pass through when disarmed like we do on CoaxCopter, SingleCopter and TriCopter. The only reason Heli don't do this is they want to be able to fly in Stabilize or Acro when disarmed.

**Comment by bnsgeyer** [comment]
> Ironically Heli doesn't do this in their own dedicated acro mode. This makes me wonder why we are compromising more semi autonomous modes when heli didn't think it was important enough in the most manual input mode....

@lthall What does this mean?  Are you saying that we always reset yaw rate in acro even before your change?  We reset the target attitudes but never the requested rates.  I missed the implications of the original change in PR #17541.  I believe we discussed this the other day that the pilot should have control of all axes in stabilize and acro.

> Landed_Ground_Idle is where we prepare the aircraft for takeoff. The presence of yaw rate from arming the aircraft can cause the aircraft to rotate on the ground, this prevents this from happening. It was the absence of this that resulted in a 50 kg multirotor rotating on the ground, wrap up a line into the propellers, and flip over while following normal take off procedures. While I understand the convenience of being able "to let those with IC engines do there checks with engine running at idle" but this has to be weighed against the inconvenience of losing an aircraft or worse, having a person injured.

I am ok with giving the heli pilot control of the yaw axis in stabilize and acro during the ground idle phase.  And we can do that without affecting multi users.

File: libraries/AC_AttitudeControl/AC_AttitudeControl.cpp [hunk 1]
```diff
@@ -887,19 +887,21 @@ void AC_AttitudeControl::reset_target_and_rate()
 }
```
Comment (bnsgeyer) [inline libraries/AC_AttitudeControl/AC_AttitudeControl.cpp:888]: reset_target_and_rates() method above should also have an option to not reset the rates so it can be used in acro flight mode for heli.

**Review by bnsgeyer — CHANGES_REQUESTED** [review]
@lthall Please modify heli acro to not reset rates in pitch, roll, and yaw axes as it did in Copter-4.0.
As for ALTHOLD and other flight modes not heli specific where pilot controls the yaw axis in ground idle, I am on the fence.  I think if any should provide the opportunity to check the yaw axis, it would be ALTHOLD and potentially POSHOLD since they have pilot control on all axes except collective.  I think it would be fine for them not to be able to check controls in ground idle in those modes.  It would come down to a procedural change for them which is much less of an impact than the code bloat to accommodate this requirement.

**Comment by lthall** [comment]
It was me that set the acro rate off too. Sorry I missed that!

I will do the same thing to the roll and pitch rate function.

File: ArduCopter/mode_althold.cpp [hunk 2]
```diff
@@ -45,7 +45,7 @@ void ModeAltHold::run()
 
     case AltHold_MotorStopped:
         attitude_control->reset_rate_controller_I_terms();
-        attitude_control->reset_yaw_target_and_rate();
+        attitude_control->reset_yaw_target_and_rate(false);
         pos_control->relax_z_controller(0.0f);   // forces throttle output to decay to zero
         break;
```
[the remark this answers is not in the document]
Comment (lthall) [reply ArduCopter/mode_althold.cpp:51]: Any Altitude hold mode is a semi automated take off and we must ensure that all controllers and actuators are conditioned correctly at the time of take off. To maintain safe operation we have to ensure that we have purged all integrator and accumulator type variables.

**Comment by lthall** [comment]
Ok, so "Ironically" the acro problem was introduced by me at the same time I introduced the rest of the problem.

To summarize the behaviour:
Heli Control checks (via the rate controllers) can be run on all four axis in Acro and Stab when in Motor_Stopped and Landed_Ground_Idle .
Heli Control checks (via the rate controllers) can be run on the roll, pitch and yaw axis in non-GPS, altitude hold based modes when in Motor_Stopped.
Other actuator frame types provide direct actuator output when disarmed in all modes OR control checks (via the rate controllers) when in Motor_Stopped in non-GPS modes.

File: ArduCopter/mode_acro_heli.cpp [hunk 3]
```diff
@@ -45,14 +45,14 @@ void ModeAcro_Heli::run()
     switch (motors->get_spool_state()) {
     case AP_Motors::SpoolState::SHUT_DOWN:
         // Motors Stopped
-        attitude_control->reset_target_and_rate();
+        attitude_control->reset_target_and_rate(true);
```
Comment (bnsgeyer) [inline ArduCopter/mode_acro_heli.cpp:48]: Shouldn't the argument read "false" since we don't want it to reset rates?

File: ArduCopter/mode_acro_heli.cpp [hunk 4]
```diff
@@ -45,14 +45,14 @@ void ModeAcro_Heli::run()
...
         attitude_control->reset_rate_controller_I_terms();
         break;
     case AP_Motors::SpoolState::GROUND_IDLE:
         // If aircraft is landed, set target heading to current and reset the integrator
         // Otherwise motors could be at ground idle for practice autorotation
         if ((motors->init_targets_on_arming() && motors->using_leaky_integrator()) || (copter.ap.land_complete && !motors->using_leaky_integrator())) {
-            attitude_control->reset_target_and_rate();
+            attitude_control->reset_target_and_rate(true);
```
Comment (bnsgeyer) [inline ArduCopter/mode_acro_heli.cpp:55]: same here

**Review by bnsgeyer — CHANGES_REQUESTED** [review]
@lthall I think you have the wrong argument for the acro modes.  In any case I think the rest is a reasonable approach.  Heli users may have to adjust their procedures to perform checks before arming for non-GPS modes using althold controller.

File: ArduCopter/mode_acro_heli.cpp (hunk 3 above)
Comment (lthall) [reply ArduCopter/mode_acro_heli.cpp]: yep, sorry

File: ArduCopter/mode_acro_heli.cpp (hunk 4 above)
Comment (lthall) [reply ArduCopter/mode_acro_heli.cpp]: yep, sorry

**Comment by Gone4Dirt** [comment]
I have flashed this to one of my helis and all is working as expected.  We can again perform control checks as following:
Acro: Roll, Pitch, Yaw, & Collective
Stabilize: Roll, Pitch, Yaw, & Collective
Alt Hold: Roll, Pitch, & Yaw
Pos Hold: Roll, Pitch, & Yaw
Drift: Roll, Pitch, & Yaw
Sport: Roll, Pitch, & Yaw
All other flight modes:  None

I think all of these make sense and are as per 4.0.7.  So I am happy.  Thanks for sorting that @lthall.

**Comment by lthall** [comment]
@Gone4Dirt That is great Matt, Thanks!!

**Review by Hwurzburg — APPROVED** [review]
Like the changes vs modes....

**Review by bnsgeyer — APPROVED** [review]
